src/DefinitionIterator.php

```diff
@@ -51,42 +53,88 @@ public function get($lookup, bool $updateContext = true)
             throw new \RuntimeException('Definition appears to contain a loop: ' . json_encode($this->lookupStack));
         }
 
-        $this->lookupStack[] = $lookup;
+        if ($definition === null) {
+            if (!$this->getRootDefinition()->has($lookup)) {
+                throw new \RuntimeException(sprintf(
+                    'No definition for %s',
+                    \is_string($lookup) || is_numeric($lookup) ? $lookup : \gettype($lookup)
+                ));
+            }
+
+            $definition    = $this->getRootDefinition()->get($lookup);
+            $updateContext = true;
+        }
 
-        $definition = $this->definition->get($lookup);
+        if ($this->context->isBuiltinValue($definition)) {
+            if ($updateContext) {
+                $this->context->set($lookup, $definition);
+            }
 
-        if ($definition === null) {
-            throw new \RuntimeException('No definition for ' . (is_scalar($lookup) ? $lookup : \gettype($lookup)));
+            return $definition;
         }
 
+        $lookup = $definition instanceof Definition ? $definition->getTreeAddress() : $lookup;
+
+        $this->lookupStack[] = $lookup;
+
         $resolver = ResolverFactory::get($definition);
 
+        // Treat $definition as an address for a different part of Definition tree
         if ($resolver === null && is_scalar($definition)) {
             $value = $this->get($definition);
 
+            if ($updateContext) {
+                $this->context->set($lookup, $value);
+            }
+
             array_pop($this->lookupStack);
 
             return $value;
         }
 
+        $value = $this->getFromResolver($lookup, $definition, $resolver);
```

Comment:
Looks like you can wrap value setting logic and remove duplicate lines.

`$value = ($resolver === null && is_scalar($definition)) ? $this->get($definition) : $this->getFromResolver($lookup, $definition, $resolver);`

Reply:
Good catch!